crates/bevy_ecs/src/observer/builder.rs

```diff
@@ -0,0 +1,239 @@
+use std::any::TypeId;
+
+use super::*;
+
+/// Builder struct for [`Observer`].
+pub struct ObserverBuilder<'w, E = ()> {
+    commands: Commands<'w, 'w>,
+    descriptor: ObserverDescriptor,
+    _marker: PhantomData<E>,
+}
+
+impl<'w, E: 'static> ObserverBuilder<'w, E> {
+    /// Constructs a new [`ObserverBuilder`].
+    pub fn new(commands: Commands<'w, 'w>) -> Self {
+        let mut descriptor = ObserverDescriptor::default();
+        let event = commands
+            .components()
+            .get_id(TypeId::of::<E>())
+            .unwrap_or_else(|| {
+                panic!(
+                    "Cannot observe event before it is registered with init_component: {}",
+                    std::any::type_name::<E>(),
+                )
+            });
+        descriptor.events.push(event);
+
+        Self {
+            commands,
+            descriptor,
+            _marker: PhantomData,
+        }
+    }
+
+    /// Constructs an [`ObserverBuilder`] with a dynamic event id.
```

Comment:
These docs need more explanation for "why might this be useful". Right now it mostly looks like a bad version of  `new`.